fms/modules/linear.py

```diff
@@ -0,0 +1,178 @@
+
+
+__type_factory_map: Mapping[str, Callable] = {}
+__type_sharding_map: Mapping[str, Callable] = {}
+
+
+def register_linear_type_to_module_map(linear_type: str, factory: Callable) -> None:
+    if linear_type in __type_factory_map:
+        raise KeyError(
+            f"Module mapping of linear type `{linear_type}` already registered"
+        )
+    __type_factory_map[linear_type] = factory
+
+
+def register_linear_type_to_sharding_map(linear_type: str, factory: Callable) -> None:
+    if linear_type in __type_sharding_map:
+        raise KeyError(
+            f"Sharding map of linear type `{linear_type}` already registered"
+        )
+    __type_sharding_map[linear_type] = factory
+
+
+def get_all_linear_type_to_sharding_maps() -> dict[str, Callable]:
+    return __type_sharding_map
+
+
+def get_linear_type(linear_config:  Optional[Mapping[str, Any]]) -> str:
+    if not linear_config:
+        return "torch_linear"
+
+    linear_type = linear_config.get("linear_type", None)
+    if not linear_type:
+        return "torch_linear"
+    if not isinstance(linear_type, str):
+        raise TypeError("linear_type in linear_config must be string")
+    if linear_type.lower() not in __type_factory_map:
+        raise ValueError(f"Unsupported linear_type `{linear_type}` in linear_config")
+
+    return linear_type.lower()
+
+
+def get_linear(
+    in_features: int,
+    out_features: int,
+    bias: bool,
+    linear_config:  Optional[Mapping[str, Any]] = None,
+) -> nn.Module:
+    linear_type = get_linear_type(linear_config)
+
+    # TODO: how to merge these calls that get different arguments?
+    if linear_type in __type_factory_map:
+        if linear_type == "torch_linear":
+            return __type_factory_map[linear_type](in_features, out_features, bias)
+        else:
+            return __type_factory_map[linear_type](
+                in_features, out_features, bias, linear_config
+            )
+    else:
+        raise KeyError(f"Unsupported linear type `{linear_type}`")
+
+
+@dataclass
+class LinearModuleShardingInfo:
+    linear_module: torch.nn.Module
+    sharding_dim: int
+    max_partitions: List[int]
+
+
+@dataclass
+class LinearParameterShardingInfo:
+    sharding_dim: int
+    shard_type: ShardType
+
+
+def shard_base_linear(
+    tensor_values: Dict[str, torch.Tensor],
+    tp_module: TPModule,
+    module_sharding_info: Dict[str, LinearModuleShardingInfo],
+    param_sharding_info: Dict[str, Dict[str, LinearParameterShardingInfo]],
+) -> None:
+    all_params = {}
+    used_keys: set[str] = set()
+
+    # Collect quantization parameters to copy on sharded module
+    tensor_device = None
+    param_count = 0
+    for module_name in module_sharding_info:
+        for param_name in param_sharding_info[module_name]:
+            if module_name not in all_params:
+                all_params[module_name] = {}
+            # TODO: reusing method '_get_sd_weight' but consider changing its name
```

Comment:
what would be a more appropriate name?

Reply:
what about `_get_sd_param`? It is used to load a general parameter, not necessarily a weight. I was hesitant to adjust this method name simply because this method is called in various places within FMS. Not a problem to rename these calls (~10) though.

Reply:
feel free to rename, although it can wait for the follow-up PR